### ISSUE TITLE
Using callbacks in a C context

### ISSUE THREAD

**Issue by restitux** [ValveSoftware/GameNetworkingSockets#217]
I think it would be helpful to add some documentation on how to get the callbacks working in a C environment. The typical approach might be to use a global variable since the callbacks don't take a `void *` (I know this is what I tried at first). The likely better approach would be to use: https://github.com/ValveSoftware/GameNetworkingSockets/blob/3e07d0165da19bd1cc165fdce7c0e022ada58c80/include/steam/isteamnetworkingsockets.h#L219 and pass some `void *` that points to some per-netconnection object with all the state your callbacks would need stored in it. 

I'm not sure if this actually documented somewhere I didn't read, but at least based on the example I didn't see any use of it.  

I could write a small example if that would be helpful.

**Comment by fletcherdvalve** [comment]
Adding examples and documentation is good.  But there's one very sharp edge here which needs to be pointed out regarding callbacks and user data.  Callbacks go into a queue at the time they happen, and the userdata from the connection is captured at the time the event happens and the callback is queued.  It's possible for an incoming connection to change state before the app is even aware of the connection's existence.  E.g.

* Incoming connection, callback is queued with userdata at default value = 0.
* Very soon after, connection changes state (perhaps client closes it, etc), callback will be queued, again with userdata at default value of 0.
* App polls callbacks.
* * The first callback is the incoming connection.  it's a new connection, so the app sets the userdata.
* * Now it gets the second callback.  It's not for a new connection, so they app might assume "all existing connections have their userdata set".  But this is not true.

Does that make sense?  Because of the need to handle this particular race condition, applications cannot solely rely on the userdata to link up the connection with the app data structures.  That association must be based on the connection handle.  (But the userdata could be used as an optimization for the very common case where this race condition did not happen.)

This race condition does not happen with incoming messages, BTW.  When you receive messages, all messages that are returned will have the user data field set based on its value at that moment, NOT the userdata that was in effect at the time the message was actually received.

**Comment by restitux** [comment]
Gotcha, that behavior is important to consider. Does this race condition still happen in the case where userData is set using the initialization options passed to functions like `CreateListenSocketIP`?

**Comment by fletcherdvalve** [comment]
Right, if you are able to set the desired user data through config vals, it will be active when the connection is created, and that would work.  Same thing applies if you pass it to the functions that create outbound connections, such as `ConnectByIPAddress`.

But with incoming connections, if you set it on the listen socket and it is inherited, that means the value will be the same for all incoming connections.
